stdlib/LinearAlgebra/src/uniformscaling.jl

```diff
@@ -141,23 +165,29 @@ function det(J::UniformScaling{T}) where T
     end
 end
 
+function tr(J::UniformScaling{T}) where T
+    if iszero(J.λ)
+        zero(T)
+    else
+        throw(ArgumentError("Trace of UniformScaling is only well-defined when λ = 0"))
+    end
+end
+
 *(J1::UniformScaling, J2::UniformScaling) = UniformScaling(J1.λ*J2.λ)
 *(B::BitArray{2}, J::UniformScaling) = *(Array(B), J::UniformScaling)
 *(J::UniformScaling, B::BitArray{2}) = *(J::UniformScaling, Array(B))
-*(A::AbstractMatrix, J::UniformScaling) = A*J.λ
+*(A::AbstractVecOrMat, J::UniformScaling) = A*J.λ
```

Comment:
Sorry my linear algebra class was not taught in English, I can't get what "the λ-fold of the identity matrix" means, could you explain it in detail?

Reply:
It's just saying that it's some identity matrix multiplied by λ.

This operation, multiplying a `vec*mat` is frequently an error.  I typically think of an identity matrix as being larger than 1x1, so I initially agreed.  But this is well-defined if J is considered to be 1x1... and we do say it automatically infers its size based upon the operation at hand.

I do think some extra consideration is still needed here, though, as multiplying `rand(5) * fill(2, 1, 1)` yields a 5×1 matrix and not a 5-element vector (as it would return as it's currently defined).

Reply:
Yes, the issue is the distinction between a simple vector ``v`` and the linear map ``x \mapsto v*x``. In the first case, `v` should be a vector as we know it. In the second, `v` should be an nx1 matrix. Take your example `rand(5) * fill(2, 1, 1)`. As it stands, it is non-sensical. Checking with `@which` what is actually done leads to

https://github.com/JuliaLang/julia/blob/d60503a00edfa19d6b13f8aba8ff3e0e4cdf9ae0/stdlib/LinearAlgebra/src/matmul.jl#L62

You see, to perform the product you first need to reshape the vector into a matrix, and only then multiplication works as expected. This and the lines above the quote are somewhat strange "convenience" functions, I would say. Anyway, if we wanted to be consistent with what we already have, we should have
```
*(A::AbstractMatrix, J::UniformScaling) = A*J.λ
*(a::AbstractVector, J::UniformScaling) = reshape(a*J.λ, length(a), 1)
```

Reply:
You'll probably enjoy the discussion in #4774 if you've not discovered that one yet.  In short, we've decided to allow vectors to participate in the linear algebra of matrices — and in many cases that means they behave as though they have an implicit trailing `x1` dimension.  This particular case is covered in https://github.com/JuliaLang/julia/pull/20423.  Sure it's implemented as a reshape, but the important part is the behavior — and this is the behavior we've decided to have.

Reply:
Thanks @mbauman, for the link. I remember reading it some long time ago. I agree it's a design decision, and I thought that behavior is desired. Would
```
*(a::AbstractVector, J::UniformScaling) = reshape(a, length(a), 1) * J
```
be then more appropriate? We would leave the multiplication handling to more generic parts of the code.

Reply:
Thanks you two! I also consider `*(a::AbstractVector, J::UniformScaling) = reshape(a, length(a), 1) * J` more reasonable